### PR TITLE
Use a table specifically for transaction trees witnesses

### LINKED ISSUE #5883: Reduce index database access time for transaction trees queries
When run against a fully populated PostgreSQL-backed participant index, queries to retrieve transaction trees are very slow.

### Example

```
explain analyze SELECT event_offset, transaction_id, node_index, participant_events.event_id, contract_id, ledger_effective_time, template_id, workflow_id, create_argument, create_signatories, create_observers, create_agreement_text, create_key_value, exercise_consuming, exercise_choice, exercise_argument, exercise_result, exercise_actors, exercise_child_event_ids, ARRAY [ '00001' ] AS event_witnesses, CASE WHEN submitter = '00001' THEN command_id ELSE '' END AS command_id FROM participant_events NATURAL JOIN ( SELECT event_id, event_witness FROM participant_event_flat_transaction_witnesses UNION SELECT event_id, event_witness FROM participant_event_witnesses_complement) AS participant_event_witnesses WHERE event_offset > '\x0000000000003a020000000000000000' AND event_offset <= '\x0000000000003a060000000000000000' AND event_witness = '00001' ORDER BY (event_offset, transaction_id, node_index) LIMIT '1000' OFFSET '0';
```

### Query Plan
```
 Limit  (cost=32851.26..32851.38 rows=50 width=1781) (actual time=1724.923..1724.929 rows=42 loops=1)
   ->  Sort  (cost=32851.26..32851.38 rows=50 width=1781) (actual time=1724.922..1724.925 rows=42 loops=1)
         Sort Key: (ROW(participant_events.event_offset, participant_events.transaction_id, participant_events.node_index))
         Sort Method: quicksort  Memory: 95kB
         ->  Merge Join  (cost=29474.53..32849.84 rows=50 width=1781) (actual time=1348.854..1724.648 rows=42 loops=1)
               Merge Cond: ((participant_event_flat_transaction_witnesses.event_id)::text = (participant_events.event_id)::text)
               ->  Unique  (cost=29368.44..30633.86 rows=168722 width=64) (actual time=1279.993..1678.244 rows=166690 loops=1)
                     ->  Sort  (cost=29368.44..29790.25 rows=168722 width=64) (actual time=1279.990..1643.946 rows=166690 loops=1)
                           Sort Key: participant_event_flat_transaction_witnesses.event_id, participant_event_flat_transaction_witnesses.event_witness
                           Sort Method: external merge  Disk: 16104kB
                           ->  Append  (cost=0.00..8374.26 rows=168722 width=64) (actual time=0.013..57.693 rows=169248 loops=1)
                                 ->  Seq Scan on participant_event_flat_transaction_witnesses  (cost=0.00..5399.59 rows=136743 width=87) (actual time=0.013..38.414 rows=137207 loops=1)
                                       Filter: ((event_witness)::text = '00001'::text)
                                       Rows Removed by Filter: 64381
                                 ->  Seq Scan on participant_event_witnesses_complement  (cost=0.00..1287.45 rows=31979 width=87) (actual time=0.006..6.562 rows=32041 loops=1)
                                       Filter: ((event_witness)::text = '00001'::text)
                                       Rows Removed by Filter: 16009
               ->  Sort  (cost=106.08..106.21 rows=50 width=1710) (actual time=5.146..5.151 rows=42 loops=1)
                     Sort Key: participant_events.event_id
                     Sort Method: quicksort  Memory: 87kB
                     ->  Index Scan using participant_events_event_offset_transaction_id_node_index_idx on participant_events  (cost=0.42..104.67 rows=50 width=1710) (actual time=3.403..4.838 rows=42 loops=1)
                           Index Cond: ((event_offset > '\x0000000000003a020000000000000000'::bytea) AND (event_offset <= '\x0000000000003a060000000000000000'::bytea))
 Planning time: 3.965 ms
 Execution time: 1725.391 ms
```

Early experiments revealed that performances can be greatly improved by:
1. pre-computing the transaction trees witnesses
2. change the indexing scheme to have a single, primary key on the `event_id, event_witness` tuple

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
This allows to not re-compute those every time using the flat transaction
witnesses and the complement.

Furthermore, witness tables indexing is redefined to a single primary key
covering both event identifier and witness party instead of two
separated indexes. This has proven to be a relevant improvement for
performance.

Fixes #5883

changelog_begin
[Ledger API Server] Significant improvements in serving transaction trees. See #5883.
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
The _longest running_ query in my tests clocks in at 11ms (on the database), versus the previous _mean_ of 121ms. Metrics point at a _mean_ execution time of 7ms (including the roundtrip between the app and the database).

File: ledger/sandbox/src/main/resources/db/migration/postgres/V31__Event_witnesses_single_table.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+insert into participant_event_transaction_tree_witnesses
+    select event_id, event_witness from participant_event_flat_transaction_witnesses
+    union
+    select event_id, event_witness from participant_event_witnesses_complement;
+
+drop index participant_event_flat_transaction_witnesses_event_id_idx;
+drop index participant_event_flat_transaction_witnesses_event_witness_idx;
+alter table participant_event_flat_transaction_witnesses add primary key (event_id, event_witness);
+
+drop table participant_event_witnesses_complement;
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/resources/db/migration/postgres/V31__Event_witnesses_single_table.sql:22]: Newline please.

Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/resources/db/migration/postgres/V31__Event_witnesses_single_table.sql:22]: I kept them close on purpose.

Comment (SamirTalwar-DA) [reply ledger/sandbox/src/main/resources/db/migration/postgres/V31__Event_witnesses_single_table.sql:22]: No, I mean a newline at the end of the file. It's missing one.

**Pull request merged**